sia/renter/rent.go

```diff
@@ -0,0 +1,159 @@
+		return
+	}
+
+	// Find a host. If the search or the negotiation is unsuccessful,
+	// hostdb.FlagHost() will be called and another host will be requested. If
+	// there is an internal error (no hosts, or an unsuccessful flagging for
+	// example), the loop will break.
+	var host components.HostEntry
+	var fileContract consensus.FileContract
+	for {
+		host, err = r.hostDB.RandomHost()
+		if err != nil {
+			return
+		}
+
+		// Fill out the contract according to the whims of the host.
+		// The contract fund: (burn * duration + price * full duration) * filesize
+		delay := consensus.BlockHeight(20)
+		contractFund := (host.Price*consensus.Currency(duration+delay) + host.Burn*consensus.Currency(duration)) * consensus.Currency(info.Size())
+		fileContract = consensus.FileContract{
+			ContractFund:       contractFund,
+			FileMerkleRoot:     merkle,
+			FileSize:           uint64(info.Size()),
+			Start:              r.state.Height() + delay,
+			End:                r.state.Height() + duration + delay,
+			ChallengeWindow:    host.Window,
+			Tolerance:          host.Tolerance,
+			ValidProofPayout:   host.Price * consensus.Currency(info.Size()) * consensus.Currency(host.Window),
+			ValidProofAddress:  host.CoinAddress,
+			MissedProofPayout:  host.Burn * consensus.Currency(info.Size()) * consensus.Currency(host.Window),
+			MissedProofAddress: consensus.CoinAddress{}, // The empty address is the burn address.
+		}
+
+		// Fund the client portion of the transaction.
+		minerFee := consensus.Currency(10) // TODO: ask wallet.
+		renterPortion := host.Price * consensus.Currency(duration+delay) * consensus.Currency(fileContract.FileSize)
+		var id string
+		id, err = r.wallet.RegisterTransaction(consensus.Transaction{})
+		if err != nil {
+			return
+		}
+		err = r.wallet.FundTransaction(id, renterPortion+minerFee)
+		if err != nil {
+			return
+		}
+		err = r.wallet.AddMinerFee(id, minerFee)
+		if err != nil {
+			return
+		}
+		err = r.wallet.AddFileContract(id, fileContract)
+		if err != nil {
+			return
+		}
+		var transaction consensus.Transaction
+		transaction, err = r.wallet.SignTransaction(id, false)
+		if err != nil {
+			return
+		}
+
+		// Negotiate the contract to the host.
+		err = host.IPAddress.Call("NegotiateContract", func(conn net.Conn) error {
+			// send contract
+			if _, err := encoding.WriteObject(conn, transaction); err != nil {
+				return err
+			}
+			// read response
+			var response string
+			if err := encoding.ReadObject(conn, &response, 128); err != nil {
+				return err
+			}
+			if response != components.AcceptContractResponse {
+				return errors.New(response)
+			}
+			// host accepted, so transmit file data
+			// (no prefix needed, since FileSize is included in the metadata)
+			_, err = io.CopyN(conn, file, info.Size())
+			return err
+		})
+		if err == nil {
+			break
+		}
+
+		fmt.Println("Problem from NegotiateContract:", err)
+		err = r.hostDB.FlagHost(host.ID)
+		if err != nil {
+			return
+		}
```

Comment:
*proposed error conventions, I'm not certain they're all good ideas.

But problems I spot with this code:

1. `err == nil` just worries me because that's confusing. I do like always having `err != nil`. I think there's a good chance of mistakes.

2.
```go 
err := // long stuff
        // long stuff
        /// long stuff
if err //
```
This makes me think there's a risk the error gets missed or misinterpreted or something.

So let's try:

```go
err = host.IPAddress.Call("NegotiateContract", handleNegotiateContract)
if err != nil {
    fmt.Println("problem")
    r.handleFailedNegotiation(host, err)
}
```